Fix denominator of roundto result

maxnikulin opened the pull request:
Multiprecision rational created from a double value may have large power of 2 denominator since fractional decimal numbers can not be represented as binary floating point numbers. It leads to failed assertion when result is compared to a value converted directly from strings. Use integer multiprecision arithmetics to round numbers to ensure proper denominator. Inspired by python gmpy2 package <https://github.com/aleaxit/gmpy/blob/3e4564ae9d/src/gmpy2_mpq_misc.c#L315>

See #2329

jwiegley commented:
Love it, once the tests pass!

tbm commented:
@maxnikulin can you take a look at the test failure.

maxnikulin commented:
Almost certainly tests are failing because I added `_dup()` call similar to `amount_t::in_place_floor()` and `amount_t::in_place_ceiling()`. Is the following behavior expected?
```c++
balance_t b;
amount_t a("$ 123.123");
b += a;
// b.to_string() == "$ 123.123"
a.in_place_roundto(2);
// b.to_string() == "$ 123.120"
```
I am surprised that changing amount after adding it to a balance affects balance value.

Do you have any idea concerning content of ledger file that is sensitive to `_dup` call during ceiling/floor/round operations? I do not like current inconsistency.

maxnikulin commented:
The following tests fail with original code and I do not like that `in_place_roundto()` affects the variable from which amount was copied. I believe, my current variant is more correct:
```c++
  amount_t a1("$ 123.123");
  amount_t a2(a1);
  a2.in_place_roundto(2);
  // Original version fails due to denominator difference with
  // error: in "balance/testRoundMinimal": check amount_t("$ 123.120") == a2 has failed [$ 123.120 != $ 123.120]
  // But why not "$ 123.12"?
  BOOST_CHECK_EQUAL(amount_t("$ 123.120"), a2);
  // Original version fails with
  // error: in "balance/testRoundMinimal": check amount_t("$ 123.123") == a1 has failed [$ 123.123 != $ 123.120]
  BOOST_CHECK_EQUAL(amount_t("$ 123.123"), a1);
```

By the way, `amount_t::roundto(int)` ignores its argument. Looks like another bug.

maxnikulin commented:
I have update the commit, `_dup()` is not called any more. I have created #2362 to track it.

I had to update another test case. I suppose `-123,40` vs. `123,50` rounding for the same absolute value is bug #1983 and this change fixes it.

https://github.com/ledger/ledger/blob/92db49d2e0d649c31b93be1e0cd0b5cde5ce57d8/doc/ledger3.texi#L8658-L8664

https://github.com/ledger/ledger/blob/92db49d2e0d649c31b93be1e0cd0b5cde5ce57d8/doc/ledger3.texi#L8430-L8435

jwiegley commented:
Your assessment of the issues around `_dup` makes sense to me, and I'd definitely like to see this addressed in a follow-up PR. Thank you for giving this so much thought.

maxnikulin commented:
> Your assessment of the issues around `_dup` makes sense to me, and I'd definitely like to see this addressed in a follow-up PR. Thank you for giving this so much thought.

I am sorry, but `round()` and `unround()` are beyond my knowledge related to ledger and these functions are not documented. I have no idea what is purpose of their magic. That is why I can not guess what `balance/testRound` is supposed to test (ledger/ledger@c5f41b8519). Unfortunately tripun who made greet job adding tests has not been active for a decade.

My questions related to the change:
1. Are you OK that I modified expectation of a unit test?
2. Do you agree with round to even implemented by this commit and thus
   ```
   build/ledger -f /dev/null eval 'roundto(0.005, 2)'
   0
   ```
   or do you prefer
   ```
   ledger -f /dev/null eval 'roundto(0.005, 2)'
   0.01
   ```
   like in #1663 and
   ```
   ledger -f /dev/null eval -- '-roundto(0.005 + 0.015 + 0.025 + 0.035 + 0.045, 2) + roundto(0.005, 2) + roundto(0.015, 2) + roundto(0.025, 2) + roundto(0.035, 2) + roundto(0.045, 2)'
   0.02
   ```
   The proposed commit gives
   ```
   build/ledger -f /dev/null eval -- '-roundto(0.005 + 0.015 + 0.025 + 0.035 + 0.045, 2) + roundto(0.005, 2) + roundto(0.015, 2) + roundto(0.025, 2) + roundto(0.035, 2) + roundto(0.045, 2)'
   0
   ```
   It should be possible to avoid `0.49999999` magic constant, but achieve 0.005 to 0.01 and -0.005 to -0.01 rounding.
3. In second patch revision I tried to add aliases to `temp` and `tempq`, but perhaps it should be reverted since `mpz_t` is a kind of trick while actual `__mpz_struct` is not a part of API.

jwiegley commented:
Amounts are infinite precision internally (stored as fractions), but they are _displayed_ at the precision observed in relation to the commodity.

`round()` causes this rounded to occur immediately and then be stored back into the fraction. For example, if the amount were `$0.3333333333333333`, this would be represented by the fraction `1/3`. It might display as `$0.33`.

`round()` would change the fraction from `1/3` to `33/100`.

`unround()` tells that particular amount to never use commodity-driven rounding logic, but to display with as much precision as is allowed by the context.

Strangely (now that I think about it), there is no way to undo the effect of `unround()`. It should probably be a method like `autoround()` or some such.

maxnikulin commented:
> `round()` causes this rounding to occur immediately

Unfortunately I can not figure out what I have missed. `amount_t::in_place_round()` just sets a flag that affects `quantity->prec` after following operations. It does not affect arithmetic.

That is why I can not get the point of the following assertions in `balance/testRound`:
```
(0.00 + $ 123.123 + EUR 456.567).rounded() ==
 0.00 + $ 123.120 + EUR 456.570
```
and the same with `in_place_round()`.

The test looks like a copy-paste of `testCeiling` or `testFloor` that is not split into `round` and `roundto`. The latter is more suitable for the used checks.

jwiegley commented:
Yeah, that test is very fishy. It is asserting something that is not generally true:
```
(x + y).round() == x.round() + y.round()
```

maxnikulin commented:
> Yeah, that test is very fishy.

Do you have any idea what test for `balance_t` may be meaningful for `rounded()`/`in_place_round()` and similar code?

It may be made valid if `roundto(2)` and `in_place_roundto(2)` are used instead. In addition it requires a fix in `amount_t::roundto(int)` however.

Out of curiosity I have tried `amount_t::in_place_truncate()` implemented as `in_place_roundto(display_precision())` from this pull request to avoid conversion to and from string. It seems no tests are failing.

maxnikulin commented:
I have rebased the commit to master and added references to #2362.

Attempts of further fixes would either depend on this pull request or would conflict with it. Feel free to cherry-pick specific chunks and close this pull requests. I am realizing that sometimes it is not easy to synchronize intervals of activity.

jwiegley commented:
So, materially, how do you think this could affect people's current data? I imagine we will see reported amounts changing if we merge this?

maxnikulin commented:
> So, materially, how do you think this could affect people's current data? I imagine we will see reported amounts changing if we merge this?

It is hard to reason for me since my attempt of using `roundto()` in a ledger file failed. I had a hope to use it as a workaround for a case close to [tbm. Rounding in ledger. Jun 24, 2014, 5:06:19 AM](https://groups.google.com/g/ledger-cli/c/m-TgILbfrwA/). I am trying to use assertions as extensive as possible to catch typos in positngs and errors in automatic transactions. Certainly behavior will change for people using `roundto()` in custom reports formats. At least in some cases output will be less confusing, like in the example in the docs (see above).

Again, if you strongly believe that halves should be rounded away from zero, not to nearest even then it can be tried.

For a while, comparison of results for some expressions evaluated by original and patched versions is the best what I can provide:
```
ledger eval "..."                    orig        fix match

# Denominator of rounded value
roundto(1.1, 1) == 1.1                  0          1     -
(roundto(1.1, 1) - 1.1)*1,000,000,000,000,000,000
\_                                   88.8          0     -
roundto(11, 0) == 11                    1          1     +
roundto(123, -1) == 120.00              1          1     +

# Rounding of halves for positive and negative values
roundto( 2.125, 2)                   2.13       2.12     -
roundto(-2.125, 2)                  -2.12      -2.12     +
roundto( 3.135, 2)                   3.14       3.14     +
roundto(-3.135, 2)                  -3.13      -3.14     -

# Accumulated error
## Alternating
0.005 + 0.015 + 0.025 + 0.035        0.08       0.08     +
roundto( 0.005, 2) + roundto( 0.015, 2) + roundto( 0.025, 2) + roundto( 0.035, 2)
\_                                    0.1       0.08     -
roundto(-0.005, 2) + roundto(-0.015, 2) + roundto(-0.025, 2) + roundto(-0.035, 2)
\_                                  -0.06      -0.08     -
## Odd
1.015 + 1.035 + 1.055 + 1.075        4.18       4.18     +
roundto( 1.015, 2) + roundto( 1.035, 2) + roundto( 1.055, 2) + roundto( 1.075, 2)
\_                                    4.2        4.2     +
roundto(-1.015, 2) + roundto(-1.035, 2) + roundto(-1.055, 2) + roundto(-1.075, 2)
\_                                  -4.16       -4.2     -
## Even
2.005 + 2.025 + 2.045 + 2.065        8.14       8.14     +
roundto( 2.005, 2) + roundto( 2.025, 2) + roundto( 2.045, 2) + roundto( 2.065, 2)
\_                                   8.16       8.12     -
roundto(-2.005, 2) + roundto(-2.025, 2) + roundto(-2.045, 2) + roundto(-2.065, 2)
\_                                  -8.12      -8.12     +

# Close to half values
roundto(1.1499999999,  1)             1.2        1.1     -
roundto(1.1499000,     1)             1.1        1.1     +
roundto(2.2499999999,  1)             2.3        2.2     -
roundto(2.2499000,     1)             2.2        2.2     +
roundto(-2.1500000001, 1)            -2.1       -2.2     -
roundto(-2.15001,      1)            -2.2       -2.2     +
roundto(-3.2500000001, 1)            -3.2       -3.3     -
roundto(-3.25001,      1)            -3.3       -3.3     +
```

jwiegley commented:
I certainly think that your numerical tests should define the standard of truth, and the implementation should measure up to them, or else what we have today is wrong.

I'm inclined to accept this Pull Request. Is there anyone who has an objection, or feels that the current behavior should in any way be preserved?

tbm commented:
> Is there anyone who has an objection 

I'm not sure how many people are following this PR. Maybe this change should be mentioned on the mailing list.

But it sounds like this will fix #1663 which I feel personally very strongly about. I think it should also fix #1765 (@maxnikulin can you check?) which is another great bug report showing the broken behaviour (`-B` not balancing to zero).

maxnikulin commented:
> > Is there anyone who has an objection
> 
> I'm not sure how many people are following this PR. Maybe this change should be mentioned on the mailing list.

This is a public pull request, so feel free to discuss it in any channel you find appropriate.

> But it sounds like this will fix #1663 which I feel personally very strongly about.

As I have already written, the expectation from the original report "3 * 1.595 = 4.785 which should imho be rounded up to 4.79" is not fulfilled. I am not sure what should be concluded from other comments to that issue.

Notice:
```
roundto( 2.125, 2)                   2.13       2.12     -
```
If you expect a different result, please, specify desired values for the cases from my previous comment and maybe add some more corner cases. If requirements are consistent then the code may be adjusted.

> I think it should also fix #1765 (@maxnikulin can you check?) which is another great bug report showing the broken behaviour (`-B` not balancing to zero).

It is a higher level issue.

This pull request is intended to address #2329, #1983, and #1925 with the following exception:
```
Error: Assertion failed: roundto(12.345, 2) == 12.35
```
however
```
ledger eval 'roundto(12.345, 2) == 12.34'
1
```

jwiegley commented:
I think our rounding behavior should not be based on historical expectations, but on numeric tables. It is very unfortunate that I never implemented such tests in order to nail down the semantics of `roundto`. So, even though this may change some past report outputs, I think we should accept this PR based on the test results it declares.

maxnikulin commented:
> @maxnikulin What is your suggestion for next steps?

@jwiegley, sorry for disturbing you, I have no suggestions yet. I am still confused, otherwise I would create another issue or even a pull request.

Yesterday I was filing another bug when I realized that I have some changes in my local clone of the repository. Some time ago I tried to eliminate floating point arithmetic from `amount_t::in_place_unreduce` in a similar way as it was done for roundto in this pull request. From my point of view, the code of unreduce resembles what you described in response to my question concerning `round()`. That is why I decided to ask if I completely missed the point of round/unround and reduce/unreduce or you described behavior of unreduce, but by mistake named it unround.

Actually I do not use reduce/unreduce, so my motivation to improve unreduce almost exhausted when realized that it may require more than a few lines of code. We may just stop discussion here.

By the way, this pull request was caused by #2329. You may close it if you think, there is no point in improving diagnostics for similar cases.
